index.js

```diff
@@ -1,39 +1,224 @@
+      return self._feeds[hkey]
+    } else {
+      // not cached
+      var store = self._storageF(FEED + hkey)
+      var feed = self._feeds[hkey] = hypercore(store, key, opts)
+      feed.once('close', function () {
+        delete self._feeds[hkey]
+      })
+      var hdkey = hcrypto.discoveryKey(id)
+      self._dkeys[hdkey] = key
+      return feed
+    }
+  } else if (/^[0-9A-Fa-f]{64}$/.test(id)) {
+    // string key
+    var key = asBuffer(id)
+    var store = self._storageF(FEED + id)
+    var hdkey = hcrypto.discoveryKey(key)
+    self._dkeys[hdkey] = key
+    var feed = self._feeds[id] = hypercore(store, key, opts)
+    feed.once('close', function () {
+      delete self._feeds[hkey]
+    })
+    return feed
+  } else if (self._lnames.hasOwnProperty(id)) {
+    // cached local name
+    return self._feeds[asHexStr(self._lnames[id])]
+  } else {
+    // not cached local name
+    var feed = hypercore(deferred(function (cb) {
+      self.fromLocalName(id, function (err, key) {
+        if (err) return cb(err)
+        if (key) {
+          // exists
+          self._lnames[id] = key
+          var hkey = asHexStr(key)
+          var hdkey = hcrypto.discoveryKey(key)
+          self._dkeys[hdkey] = key
+          self._feeds[hkey] = feed
+          cb(null, self._storageF(FEED + hkey))
+        } else {
+          // does not exist
+          cb(null, new Error('feed not found'))
+        }
+      })
+    }), opts)
+    feed.once('close', function () {
+      delete self._feeds[hkey]
+    })
+    return feed
+  }
 }
 
 // Returns boolean true/false if core is open.
-CoreStore.prototype.isOpen = function (id, cb) {
+Storage.prototype.isOpen = function (id, cb) {
+  return this._feeds.hasOwnProperty(asHexStr(id))
 }
 
 // Unload a hypercore.
-CoreStore.prototype.close = function (id, cb) {
+Storage.prototype.close = function (id, cb) {
+  var hkey = asHexStr(id)
+  if (!this._feeds.hasOwnProperty(hkey)) {
+    return nextTick(cb, new Error('feed not loaded'))
+  }
+  this._feeds[hkey].close(cb)
+  delete this._feeds[hkey]
 }
 
 // Close all hypercores.
-CoreStore.prototype.closeAll = function (cb) {
+Storage.prototype.closeAll = function (cb) {
+  var self = this
+  var pending = 1, finished = false
+  Object.keys(self._feeds).forEach(function (key) {
+    pending++
+    self._feeds[key].close(function (err) {
+      if (!finished && err) {
+        finished = true
+        cb(err)
+      } else if (!finished && --pending === 0) {
+        finished = true
+        cb()
+      }
+    })
+  })
+  if (--pending === 0 && !finished) {
+    finished = true
+    cb()
+  }
 }
 
 // Unload (if necessary) and delete a hypercore.
-CoreStore.prototype.delete = function (id, cb) {
+Storage.prototype.delete = function (id, cb) {
+  var hkey = asHexStr(id)
+  if (this._feeds.hasOwnProperty(hkey)) {
+    this._feeds[hkey].close(function (err) {
+      if (err) return cb(err)
+      self._delete(FEED + hkey, cb)
+    })
+  }
+  delete this._feeds[hkey]
```

Comment:
I would also expect this to also remove its entries from `_dkeys` and `_lnames`.